### PR TITLE
Fix truncation of 64bit integer constants on 32bit AIX

### PR DESCRIPTION
**Pull request by slontis**
If ULL is not specified for a uint64 constant value on 32bit AIX then it is truncated to 32 bits. A few tests as well as some newer algs (blake2b, keccak, ec_curve448) were causing quite a few test failures.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/blake2/blake2b.c [hunk 1]
```diff
@@ -23,10 +23,10 @@
...
 static const uint64_t blake2b_IV[8] =
 {
-    0x6a09e667f3bcc908U, 0xbb67ae8584caa73bU,
-    0x3c6ef372fe94f82bU, 0xa54ff53a5f1d36f1U,
-    0x510e527fade682d1U, 0x9b05688c2b3e6c1fU,
-    0x1f83d9abfb41bd6bU, 0x5be0cd19137e2179U
+    0x6a09e667f3bcc908uLL, 0xbb67ae8584caa73buLL,
+    0x3c6ef372fe94f82buLL, 0xa54ff53a5f1d36f1uLL,
+    0x510e527fade682d1uLL, 0x9b05688c2b3e6c1fuLL,
+    0x1f83d9abfb41bd6buLL, 0x5be0cd19137e2179uLL
```
Comment (bernd-edlinger) [inline crypto/blake2/blake2b.c]: your compiler is not C99 confromant,
see 6.4.4.1 Integer constants
in http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf
U means choose the smallest type from "unsigned int, unsigned long int, unsigned long long int"
where the value of the integer constant fits.

Comment (slontis) [reply crypto/blake2/blake2b.c]: Unfortunately on 32 bit AIX that truncates the value, and it needs to work on this platform
(Even if it needs to be done via a macro)

Comment (levitte) [reply crypto/blake2/blake2b.c]: Also, please note that C90 is our lower limit, so C99 compliance is indeed not guaranteed.

Comment (bernd-edlinger) [reply crypto/blake2/blake2b.c]: long long is a C99 feature, and that needs to be supported.

Comment (levitte) [reply crypto/blake2/blake2b.c]: We sometimes cheat.  `long long` was supported in most compilers before it entered the C standard.  I do believe this goes for AIX compilers, among others.

So yeah, we occasionally encounter the compiler that doesn't quite fullfil the C99 requirements even though it supports C99 types, and we sometimes need to adapt.  Adding `LL` *does no harm*, and is exactly to the point in this case.

Comment (bernd-edlinger) [reply crypto/blake2/blake2b.c]: Is there at least a warning when the compiler does this?

File: crypto/ec/curve448/curve448_tables.c [hunk 2]
```diff
@@ -16,325 +16,325 @@
...
     {
         {{
-            {FIELD_LITERAL(0x00cc3b062366f4cc,0x003d6e34e314aa3c,0x00d51c0a7521774d,0x0094e060eec6ab8b,0x00d21291b4d80082,0x00befed12b55ef1e,0x00c3dd2df5c94518,0x00e0a7b112b8d4e6)},
-            {FIELD_LITERAL(0x0019eb5608d8723a,0x00d1bab52fb3aedb,0x00270a7311ebc90c,0x0037c12b91be7f13,0x005be16cd8b5c704,0x003e181acda888e1,0x00bc1f00fc3fc6d0,0x00d3839bfa319e20)},
-            {FIELD_LITERAL(0x003caeb88611909f,0x00ea8b378c4df3d4,0x00b3295b95a5a19a,0x00a65f97514bdfb5,0x00b39efba743cab1,0x0016ba98b862fd2d,0x0001508812ee71d7,0x000a75740eea114a)},
+            {FIELD_LITERAL(0x00cc3b062366f4ccuLL,0x003d6e34e314aa3cuLL,0x00d51c0a7521774duLL,0x0094e060eec6ab8buLL,0x00d21291b4d80082uLL,0x00befed12b55ef1euLL,0x00c3dd2df5c94518uLL,0x00e0a7b112b8d4e6uLL)},
+            {FIELD_LITERAL(0x0019eb5608d8723auLL,0x00d1bab52fb3aedbuLL,0x00270a7311ebc90cuLL,0x0037c12b91be7f13uLL,0x005be16cd8b5c704uLL,0x003e181acda888e1uLL,0x00bc1f00fc3fc6d0uLL,0x00d3839bfa319e20uLL)},
+            {FIELD_LITERAL(0x003caeb88611909fuLL,0x00ea8b378c4df3d4uLL,0x00b3295b95a5a19auLL,0x00a65f97514bdfb5uLL,0x00b39efba743cab1uLL,0x0016ba98b862fd2duLL,0x0001508812ee71d7uLL,0x000a75740eea114auLL)},
```
Comment (bernd-edlinger) [inline crypto/ec/curve448/curve448_tables.c]: the line length is already too long, but it is difficult to see, if the values are changed or not.
that should probably done in the FIELD_LITERAL macro.

File: crypto/ec/curve448/scalar.c [hunk 3]
```diff
@@ -14,24 +14,24 @@
...
     {
         {
-            SC_LIMB(0x2378c292ab5844f3), SC_LIMB(0x216cc2728dc58f55),
-            SC_LIMB(0xc44edb49aed63690), SC_LIMB(0xffffffff7cca23e9),
-            SC_LIMB(0xffffffffffffffff), SC_LIMB(0xffffffffffffffff),
-            SC_LIMB(0x3fffffffffffffff)
+            SC_LIMB(0x2378c292ab5844f3ULL), SC_LIMB(0x216cc2728dc58f55ULL),
+            SC_LIMB(0xc44edb49aed63690ULL), SC_LIMB(0xffffffff7cca23e9ULL),
+            SC_LIMB(0xffffffffffffffffULL), SC_LIMB(0xffffffffffffffffULL),
+            SC_LIMB(0x3fffffffffffffffULL)
```
Comment (bernd-edlinger) [inline crypto/ec/curve448/scalar.c]: likewise here, I think that should be done inside the SC_LIMB macro.

File: crypto/ec/curve448/curve448_tables.c (hunk 2 above)
Comment (levitte) [reply crypto/ec/curve448/curve448_tables.c]: Or break lines:

``` C
            {FIELD_LITERAL(0x003caeb88611909fuLL,0x00ea8b378c4df3d4uLL,
                           0x00b3295b95a5a19auLL,0x00a65f97514bdfb5uLL,
                           0x00b39efba743cab1uLL,0x0016ba98b862fd2duLL,
                           0x0001508812ee71d7uLL,0x000a75740eea114auLL)},
```

Comment (levitte) [reply crypto/ec/curve448/curve448_tables.c]: Not sure that it's such a good idea to hide the size suffix like that... that makes for harder to read code.

File: crypto/blake2/blake2b.c (hunk 1 above)
Comment (kaduk) [reply crypto/blake2/blake2b.c]: Given that we're now like 20 years past C99, maybe we could think about moving the baseline from "C90 with some additional (C99) features" to "C99 but we tolerate some missing features"

Comment (kroeckx) [reply crypto/blake2/blake2b.c]: Visual studio does not do C99 ...

Comment (kaduk) [reply crypto/blake2/blake2b.c]: I am well aware of that.  We could document the bits they don't do in our exceptions list (and, they have been getting better recently with the latest C++ support, IIUC).

Comment (bernd-edlinger) [reply crypto/blake2/blake2b.c]: ack.

Comment (levitte) [reply crypto/blake2/blake2b.c]: C90 is still a baseline in some sectors, or so I'm told, and very hard to move away from.

Comment (slontis) [reply crypto/blake2/blake2b.c]: The compiler spits out a truncation warning.

File: crypto/ec/curve448/curve448_tables.c (hunk 2 above)
Comment (slontis) [reply crypto/ec/curve448/curve448_tables.c]: I did think of doing that, and then thought it was probably not a good idea. I know on a previous commit in another PR someone didnt like using my use of the string tokeniser.

Comment (slontis) [reply crypto/ec/curve448/curve448_tables.c]: Changed to use tokeniser

**Comment by levitte** [comment]
So with the added token in macros, we now have cases where you do have to add `ULL` and others where you don't. That makes for inconsistent code. I would much prefer an explicit `ULL` everywhere needed.

**Comment by bernd-edlinger** [comment]
no objections, I was just looking at the vast number of edits in this particular file,
but when the person doing the review uses some scripts to verify the correctness
that is of course also fine.

**Comment by slontis** [comment]
The macros no longer add ULL.
Curve448_tables.c was reformatted using a python script. The tests all pass. Manually editing any value in this table actually causes tests to fail in evp and ssl_test_new (so it should all be good).

**Comment by slontis** [comment]
ping : Can I get this one merged please.

**Comment by mattcaswell** [comment]
Is this for backport to 1.1.1? My approval counts for that too.

**Comment by levitte** [comment]
I'll try to cherry-pick

**Comment by levitte** [comment]
Merged.

master:
98f29466dc1ed7f80b9b8750309a41b5a1150d25 fix truncation of integers on 32bit AIX

1.1.1:
191570d0b94c88fa95b3c223456863448d697fce fix truncation of integers on 32bit AIX

**Pull request closed without merging**